File: src/sphere.rs
use crate::hittable::{HitRecord, Hittable};
use crate::ray::Ray;
use crate::{Material, Point3};
use std::cell::RefCell;
use std::rc::Rc;

pub struct Sphere {
    pub center: Point3,
    pub radius: f64,
    pub mat_ptr: Option<Rc<RefCell<dyn Material>>>,
}

impl Sphere {
    pub fn new(center: &Point3, radius: f64, material: Option<Rc<RefCell<dyn Material>>>) -> Self
    where
        Self: Sized,
    {
        Sphere {
            center: center.clone(),
            radius,
            mat_ptr: material,
        }
    }
}

impl Hittable for Sphere {
    fn hit(&self, ray: &Ray, t_min: f64, t_max: f64, rec: &mut HitRecord) -> bool {
        let oc = ray.origin - self.center;
        let a = ray.direction.length_squared();
        let half_b = oc.dot(&ray.direction);
        let c = oc.length_squared() - self.radius * self.radius;
        let discriminant = half_b * half_b - a * c;

        if discriminant < 0.0 {
            return false;
        }

        let sqrtd = discriminant.sqrt();

        // Find the nearest root that lies in the acceptable range.
        let mut root = (-half_b - sqrtd) / a;
        if root < t_min || t_max < root {
            root = (-half_b + sqrtd) / a;
            if root < t_min || t_max < root {
                return false;
            }
        }

        rec.t = root;
        rec.p = ray.at(rec.t);
        let outward_normal = (rec.p - self.center) / self.radius;
        rec.set_face_normal(ray, &outward_normal);
        rec.mat_ptr = self.mat_ptr.clone();

        true
    }
}

File: tj/sphere.py
import taichi as ti
from taichi.math import *

from ray import Ray
from hittable import Hittable, HitRecord
from material import Material


@ti.dataclass
class Sphere(Hittable):
    center: vec3
    radius: float
    mat: Material

    def __init__(self, center: vec3, radius: float, mat: Material):
        super(Sphere, self).__init__(2)
        self.center = center
        self.radius = radius
        self.mat = mat

    @ti.func
    def hit(self, ray: Ray, t_min: float, t_max: float) -> (bool, HitRecord):
        hit = True
        oc = ray.origin - self.center
        a = dot(ray.direction, ray.direction)
        half_b = dot(oc, ray.direction)
        c = dot(oc, oc) - self.radius * self.radius
        discriminant = half_b * half_b - a * c
        if discriminant < 0:
            hit = False
        sqrtd = ti.sqrt(discriminant)
        root = (-half_b - sqrtd) / a
        if root < t_min or t_max < root:
            root = (-half_b + sqrtd) / a
            if root < t_min or t_max < root:
                hit = False
        rec = HitRecord()
        if hit:
            rec.t = root
            rec.p = ray.at(rec.t)
            outward_normal = (rec.p - self.center) / self.radius
            rec.set_face_normal(ray, outward_normal)
            rec.mat = self.mat

        return hit, rec
